core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -1655,6 +1655,34 @@ public static String concatMulti(String... args) {
     return String.join("", args);
   }
 
+  /** SQL {@code CONCAT(obj0, obj1, obj2, ...)} function. */
+  public static String concatMultiObjects(Object... args) {
```

Comment:
I am afraid this will cause problems for more general programs.
The Java code generated by the enumerable implementation is relatively strongly typed, and it will expect ByteArray for the result of some calls. Please add some quidem tests to check.

Reply:
OK, it seems to be different from what I understand. Maybe I need to sort out these issues again. I will set this PR as a draft first, and reopen it after all the tests are completed.